lls/cache.c

```diff
@@ -357,33 +356,57 @@ xmit_icmp_ping_reply(struct gatekeeper_if *iface, struct rte_mbuf *pkt)
 	 * According to RFC 792 page 13: to form an echo reply
 	 * message, the source and destination addresses are simply reversed,
 	 * the type code changed to 0, and the checksum recomputed.
+	 *
+	 * According to RFC 3021: the 255.255.255.255 IP broadcast address
+	 * MUST be used for broadcast Address Mask Replies in point-to-point
+	 * links with 31-bit subnet masks. In other cases, the form
+	 * {<Network-number>, -1} should be used.
 	 */
 	icmp_ipv4 = (struct rte_ipv4_hdr *)pkt_out_skip_l2(iface, icmp_eth);
+	icmph = (struct rte_icmp_hdr *)ipv4_skip_exthdr(icmp_ipv4);
 
 	icmp_ipv4->time_to_live = IP_DEFTTL;
-	ip_addr_tmp = icmp_ipv4->src_addr;
-	icmp_ipv4->src_addr = icmp_ipv4->dst_addr;
-	icmp_ipv4->dst_addr = ip_addr_tmp;
+	if (icmph->icmp_type == ICMP_ECHO_REQUEST_TYPE) {
+		uint32_t ip_addr_tmp = icmp_ipv4->src_addr;
+		icmp_ipv4->src_addr = icmp_ipv4->dst_addr;
+		icmp_ipv4->dst_addr = ip_addr_tmp;
+	} else if (icmph->icmp_type == ICMP_MASK_REQUEST_TYPE) {
+		icmp_ipv4->src_addr = icmp_ipv4->dst_addr;
+		if (iface->ip4_addr_plen == 31)
+			icmp_ipv4->dst_addr = (uint32_t)-1;
+		else {
+			icmp_ipv4->dst_addr = (iface->ip4_addr.s_addr |
+					(~iface->ip4_mask.s_addr));
+		}
+	}
 
 	pkt->l2_len = iface->l2_len_out;
 	pkt->l3_len = ipv4_hdr_len(icmp_ipv4);
 	set_ipv4_checksum(iface, pkt, icmp_ipv4);
 
-	/*
-	 * According to RFC 792 page 14:
-	 *
-	 * (1) the data received in the echo message must be returned in
-	 * the echo reply message.
-	 *
-	 * (2) the identifier and sequence number may be used by
-	 * the echo sender to aid in matching the replies with
-	 * the echo requests.
-	 *
-	 * So, we keep these fields unmodified.
-	 */
-	icmph = (struct rte_icmp_hdr *)ipv4_skip_exthdr(icmp_ipv4);
-	icmph->icmp_type = ICMP_ECHO_REPLY_TYPE;
-	icmph->icmp_code = ICMP_ECHO_REPLY_CODE;
+	if (icmph->icmp_type == ICMP_ECHO_REQUEST_TYPE) {
+		/*
+		 * According to RFC 792 page 14:
+		 *
+		 * (1) the data received in the echo message must be returned in
+		 * the echo reply message.
+		 *
+		 * (2) the identifier and sequence number may be used by
+		 * the echo sender to aid in matching the replies with
+		 * the echo requests.
+		 *
+		 * So, we keep these fields unmodified.
+		 */
+		icmph->icmp_type = ICMP_ECHO_REPLY_TYPE;
+		icmph->icmp_code = ICMP_ECHO_REPLY_CODE;
+	} else if (icmph->icmp_type == ICMP_MASK_REQUEST_TYPE) {
```

Comment:
Let's add `likely()` to the test `(icmph->icmp_type == ICMP_MASK_REQUEST_TYPE)`.